[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::readMessage()

aperezdc opened the pull request:
#### bb3b3f72e7e2b20c13958c615956be41e893c6e5
<pre>
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::readMessage()
<a href="https://bugs.webkit.org/show_bug.cgi?id=302649">https://bugs.webkit.org/show_bug.cgi?id=302649</a>

Reviewed by Michael Catanzaro.

Use spans over the received message buffer to parse the different
elements of the message, advancing over the input idiomatically using
consumeAndReinterpretCastTo(), consumeSpan(), and skip().

Note that the message body size is always extended from uint32_t to a
size_t, and that the value is never modified and therefore it does not
need to be Checked&lt;size_t&gt; because there is no arithmetic performed on
it.

While at it, check and detect for messages missing the message name
delimiter, too short messages, or messages larger than 512 MiB (assuming
that no message will ever need to be that big). When an invalid message
is detected, use the new didReceiveInvalidMessage() helper to report it,
close the connection, and discard the read buffer.

* Source/WTF/wtf/glib/SocketConnection.cpp:
(WTF::SocketConnection::didReceiveInvalidMessage): Added.
(WTF::SocketConnection::readMessage): Modernize to use spans and add
additional sanity checks.
* Source/WTF/wtf/glib/SocketConnection.h: Added didReceiveInvalidMessage
declaration.

Canonical link: <a href="https://commits.webkit.org/303205@main">https://commits.webkit.org/303205@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c72806ba5ca21f159219cf11613126acadbcbe1a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131689 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/4107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/3943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/139198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134635 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/4107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/4107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/82420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/123752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/4107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/141844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/130195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/3849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/141844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/3930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/3943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/141844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20474 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/3903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/42689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/163161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/163161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/3864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/glib/SocketConnection.cpp:
```diff
@@ -109,44 +109,37 @@ static inline bool messageIsByteSwapped(MessageFlags flags)
+    const MessageFlags flags = consumeAndReinterpretCastTo<MessageFlags>(messageData);
+    auto nullIndex = find(messageData, '\0');
+    if (nullIndex == notFound) {
         ASSERT_NOT_REACHED();
```
Meh, thing is this is _expected_ to be missing if the peer sends malformed data. Arguably it's useful to flag problems in a debug build, but... I still don't like asserting on expected conditions. It's just nicer to reserve asserts for bugs only. So I would omit both this assert, and the next one too. Instead of `ASSERT(messageData.front() == '\0');` I would do `if (messageData.front() != '\0') return false;`

aperezdc replied:
We can remove this assert. But the second I would keep it: if we found the null string terminator, after consuming the message name we _must_ always have a null as the first item.

aperezdc replied:
After sleeping on this, I think you are wrong. It is always WebKit creating those messages, therefore must always be correct. Of course we need to handle short reads, and return `false` in that case to make the code continue reading until the end of the message.

OTOH, if the peer sends malformed data, we need to assume that it has been compromised and (potentially) acting with malicious intent. If we get a malformed message _and_ we make `::readMessage()` return `false`, then the code will be forever stuck trying to read more from the socket&mdash;we don't want that to happen.

aperezdc replied:
Well, scratch that, we do not want to kill ourselves: instead I think the best we can do is log the issue (also done indirectly by `CHECK_MESSAGE` when a connection client gets its `didReceiveInvalidMessage` function), close the connection, and discard the read buffer.

Unfortunately in this case we cannot ask the remote peer to shut down, and it may not even be on the same host (_Remote_ Web Inspector, after all!) so we cannot `kill()` the process after retrieving its PID using `SO_PEERCRED`.

aperezdc replied:
Done, now we close the connection on invalid messages.

mcatanzaro reviewed (commented):
Normally I would click the Approve button now, because this looks good to me. But you told me on Matrix that this is broken and you are still debugging it, so let's wait for the result of that investigation.

mcatanzaro commented on Source/WTF/wtf/glib/SocketConnection.cpp:
```diff
@@ -109,44 +109,36 @@ static inline bool messageIsByteSwapped(MessageFlags flags)
-    auto messageSize = sizeof(uint32_t) + sizeof(MessageFlags) + bodySize;
+    auto messageData = m_readBuffer.span();
+    const size_t bodySize = ntohl(consumeAndReinterpretCastTo<uint32_t>(messageData));
+    const size_t messageSize = sizeof(uint32_t) + sizeof(MessageFlags) + bodySize;
```
I would use `Checked<size_t>` here unless you've squinted very carefully at the rest of this function. It's clear enough that these two expressions will never overflow, but this has impacts on the rest of this function as well. E.g. `m_readBuffer.size() - messageSize.value()` was previously checked, but no longer is.

aperezdc replied:
Once you obtain the value with `messageSize.value()` the arithmetic is no longer checked.

It is obvious that `m_readBuffer.size() - messageSize` cannot overflow, because the `if` condition is _literally_ checking that the buffer is larger than the message size, just two lines above:

```cpp
    if (m_readBuffer.size() > messageSize) {
        memmoveSpan(m_readBuffer.mutableSpan(), m_readBuffer.subspan(messageSize));
        m_readBuffer.shrink(m_readBuffer.size() - messageSize);
    } else
        m_readBuffer.shrink(0);
```

aperezdc replied:
...but your comment made me look again at how `parametersSize` gets calculated:

```cpp
const auto parametersSize = bodySize - messageName.length() - 1;
```

We get `bodySize` from the message header. In _theory_ in that case we would never be able to find the null-terminator for the message name, but we are searching in a span that covers up to the end of `m_readBuffer`, which can have data for more than one message so _in practice_ it is possible to get a short `bodySize` and end up picking the message name from a follow-up message. I am going to harden this by adding two additional bits:

- Check that the `bodySize >= 2`: at least one character for the message name, and the null-terminator&mdash;and disallow an empty message name.
- Once we have extracted the `bodySize` value, use a sub-span to ensure that `messageData` covers strictly one message.
  - With this change applied, we can remove the calculation of `parametersSize`, because after we have consumed the header and message name, the size of the span will be exactly `parametersSize`.

aperezdc replied:
Done, and this change made the code easier to follow.

csaavedra commented on Source/WTF/wtf/glib/SocketConnection.cpp:
```diff
@@ -109,44 +118,52 @@ static inline bool messageIsByteSwapped(MessageFlags flags)
+    // name (an empty name is invalid), and a null terminator at the end of the name.
+    static auto constexpr MinimumMessageBodySize = 2;
+    if (bodySize < MinimumMessageBodySize)
+        return didReceiveInvalidMessage("message body too small");
```
No strong feelings but this could be of the form:

```cpp
CHECK_MESSAGE(bodySize > MinimumMessageBody, "message body too small");
```

With 
```cpp
#define CHECK_MESSAGE(condition, message) if(!condition) [[unlikely]] return didReceiveInvalidMessage(message);
```

or something like that to avoid repetition.

aperezdc replied:
Good point. Also people grepping for uses of `MESSAGE_CHECK()` macros to see where we have IPC checks will get results from this file.

mcatanzaro reviewed (approved):
Fixing `SocketConnection::sendMessage` would be another tempting future improvement.

aperezdc committed:
[GLib] Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in SocketConnection::readMessage()
https://bugs.webkit.org/show_bug.cgi?id=302649

Reviewed by Michael Catanzaro.

Use spans over the received message buffer to parse the different
elements of the message, advancing over the input idiomatically using
consumeAndReinterpretCastTo(), consumeSpan(), and skip().

Note that the message body size is always extended from uint32_t to a
size_t, and that the value is never modified and therefore it does not
need to be Checked<size_t> because there is no arithmetic performed on
it.

While at it, check and detect for messages missing the message name
delimiter, too short messages, or messages larger than 512 MiB (assuming
that no message will ever need to be that big). When an invalid message
is detected, use the new didReceiveInvalidMessage() helper to report it,
close the connection, and discard the read buffer.

* Source/WTF/wtf/glib/SocketConnection.cpp:
(WTF::SocketConnection::didReceiveInvalidMessage): Added.
(WTF::SocketConnection::readMessage): Modernize to use spans and add
additional sanity checks.
* Source/WTF/wtf/glib/SocketConnection.h: Added didReceiveInvalidMessage
declaration.

Canonical link: https://commits.webkit.org/303205@main
